### DISCUSSION TITLE
Possible bug! Investment optimization results values plugged as fix investment and becomes unfeasible

### DISCUSSION THREAD

**Discussion by fbozzolo** [spine-tools/SpineOpt.jl#624] [Support]
Hi all,

I have a system in which I have several investment decisions happening at the same time. Once I optimize the system, I take the optimized values of investment and add them as "fixed investment" (either units or storage). The system appears to be unfeasible once I do that. 
Two other important things are:
1) The system works when I use a shorter 5 hours problem (rather than the 1 year target problem).
2) If I add the optimized values of investment as a "number of units" instead of "fixed investment" it works fine.

This issue was discussed during our last new users meeting and I was suggested to try to change the tolerance and/or the solver. I did it and tried with HiGHG, Clp, GLPK, and Cbc, in both mip and lp solver. The problem remains to appear as unfeasible. Any help will be appreciated. I could send the file if someone wants to take a look at it.

Thanks in advance.

**Comment by DillonJ** [comment]
You would, for example, need to fix units_invested_available = 1 and have candidate units = 1 also. Make sure it is not fix_units_invested that you are using (this is the point in time investment decision whereas units_invested_available represents the number of already invested-in units that are available at a point in time.

Also, what is the issue with using number_of_units if this achieves what you want? Although it would be good to understand the precise source of your infeasibility.

**Reply by fbozzolo** [reply]
Thanks for the help. Unfortunately using fix units_invested_available brought another error (attached). Any clue of what happened?
Besides, my issue with using number_of_units is that it does not consider the cost associated with the investment of those units, and even though I could do it manually I would like to get that number straight from the model.
In terms of the infeasibility produced by using fix_cost_ investment I've also attached a screenshot of the console showing the infeasibility.


This is the fix units_invested_available error:

<img width="648" alt="fix cost investment avaialable error 3" src="https://user-images.githubusercontent.com/107572164/234560295-1ae41e1d-c6d9-4d9d-a303-2671a69f0356.png">

Infeasibility of fix units_invested:
<img width="649" alt="Infeasibility" src="https://user-images.githubusercontent.com/107572164/234566980-82285742-5406-4fc8-a024-6c66e466a0cf.png">

**Comment by fbozzolo** [comment]
[Data Store 1 - Copy.zip](https://github.com/spine-tools/SpineOpt.jl/files/11374956/Data.Store.1.-.Copy.zip)

This is the database of a minimum case example that shows the issue. If you run this database through a "run spineopt" block, it should show you the infeasibility I mentioned. If you change the windfarm "fix_units_investment" for "fix_units_investment_available" it will show the error I showed you before. Finally, if you just erase "fix_units_investment" from the windfarm, it should run without issues.

**Reply by mihlema** [reply]
Thanks @fbozzolo for the minimal working example. I'm looking into this now, I'll keep you posted!

**Comment by mihlema** [comment]
Hi all,
so I explored a bit on this. @fbozzolo, I'm not getting the same error as you, but I am also getting an error (might be due to different SpineOpt/SpineInterface versions, recommended to pull the latest, and to import the latest template on to your database, I think).
When using the `fix_units_invested_available`, I'm getting an error from the preprocessing:
```
Preprocessing data structure...ERROR: MethodError: Cannot `convert` an object of type SpineInterface.TimeSeries{Float64} to an object of type SpineInterface.AbstractParameterValue
``` 
from the `generate_benders_structure()` function. The underlying problem seems to be assigning fixed values for benders decomposition, which seems to be broken at this point, see line 772 in ` preprocess_data_structure.jl`  :
``` 
            obj = ent[obj_cls.name]
            fix_value = fix_param(; ent...)
            if fix_value !== nothing
                add_object_parameter_values!(obj_cls, Dict(obj => Dict(starting_param_name => fix_value)))
            end
```
@manuelma @DillonJ could this be related to some changes in SpineInterface possibly?

@fbozzolo a (very dirty, not recommded) quick fix would be to update both SpineInterface and SpineOpt, load the most recent template, and comment ouy line 773 `add_object_parameter_values!(obj_cls, Dict(obj => Dict(starting_param_name => fix_value)))` in the source code. But much rather I'd recommend for a proper fix on this.

**Reply by manuelma** [reply]
I will take a look, thanks very much for the analysis.

**Reply by fbozzolo** [reply]
Thanks for the advice, I'll update my SpineOpt, Interface and template versions.

**Reply by fbozzolo** [reply]
@mihlema I have updated everything and I got the same error as you (I guess that's something good).

**Reply by manuelma** [reply]
Will fix

**Comment by clizbe** [comment]
@manuelma Has this been fixed or should we make an issue?

**Comment by DillonJ** [comment]
@fbozzolo a relatively new way to do what you are trying to do is to use the `investment_group` functionality. You would need to create a new `investment_group` and make your `unit` part of it by defining the `unit__investment_group`... then define the `minimum_entities_invested_available` parameter for the `investment_group` to define a minimum level of investment required.

Edit: see more here: https://spine-tools.github.io/SpineOpt.jl/latest/concept_reference/Object%20Classes/#investment_group and here: https://spine-tools.github.io/SpineOpt.jl/latest/concept_reference/Parameters/#minimum_entities_invested_available
